pkg/operator/controller/controller_service_monitor.go

```diff
@@ -0,0 +1,112 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+
+	operatorclient "github.com/openshift/cluster-dns-operator/pkg/operator/client"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (r *reconciler) ensureServiceMonitor(dns *operatorv1.DNS, svc *corev1.Service, daemonsetRef metav1.OwnerReference) (*unstructured.Unstructured, error) {
+	desired := desiredServiceMonitor(dns, svc, daemonsetRef)
+
+	current, err := r.currentServiceMonitor(dns)
+	if err != nil {
+		return nil, err
+	}
+
+	if desired != nil && current == nil {
+		if err := r.client.Create(context.TODO(), desired); err != nil {
+			return nil, fmt.Errorf("failed to create servicemonitor %s/%s: %v", desired.GetNamespace(), desired.GetName(), err)
+		}
+		logrus.Infof("created servicemonitor %s/%s", desired.GetNamespace(), desired.GetName())
+		return desired, nil
+	}
+	return current, nil
+}
+
+func serviceMonitorName(dns *operatorv1.DNS) types.NamespacedName {
```

Comment:
Should this func go in `pkg/operator/controller/names.go`?

Reply:
fixed.